### PR TITLE
bv2integer is now told the width

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 5ab57728848d7469625d9f958179d89ef7d177be).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86926808

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 5eca6853a8edf3e4cf1094c29a218900b9537b71).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87137970

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: c89578a30259a6e44392a327ef3598ae9be34133).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87172948
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 8314f36d651629e81b0e608f14bd0cedf99e8c18).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87854314

File: jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp [hunk 1]
```diff
@@ -86,7 +86,9 @@ static const std::string get_thread_block_identifier(
 {
   PRECONDITION(f_code.arguments().size() == 1);
   const exprt &expr = f_code.arguments()[0];
-  mp_integer lbl_id = bv2integer(expr.op0().get_string(ID_value), false);
+  mp_integer lbl_id;
+  if(to_integer(to_constant_expr(expr.op0()), lbl_id))
+    CHECK_RETURN(false);
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp]:
```suggestion
mp_integer lbl_id = numeric_cast_v<mp_integer>(expr.op0());
```

File: src/ansi-c/expr2c.cpp [hunk 2]
```diff
@@ -1811,9 +1811,10 @@ std::string expr2ct::convert_constant(
     if(c_enum_type.id()!=ID_c_enum)
       return convert_norep(src, precedence);
 
-    bool is_signed=c_enum_type.subtype().id()==ID_signedbv;
+    const bool is_signed = c_enum_type.subtype().id() == ID_signedbv;
```
Comment (tautschnig) [inline src/ansi-c/expr2c.cpp:1814]: Nit pick: Unrelated change?

File: src/goto-programs/goto_convert.cpp [hunk 3]
```diff
@@ -1830,8 +1830,11 @@ bool goto_convertt::get_string_constant(
       forall_operands(it, index_op)
         if(it->is_constant())
         {
-          unsigned long i = integer2ulong(
-            bv2integer(id2string(to_constant_expr(*it).get_value()), true));
+          mp_integer int_value;
+          if(to_integer(to_constant_expr(*it), int_value))
+            return true;
+
+          unsigned long i = integer2ulong(int_value);
```
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]:
```suggestion
          const auto i = numeric_cast<std::size_t>(*it);
          if(!i.has_value())
            return true;
```
and then use `*i` below.

File: src/goto-programs/goto_trace.cpp [hunk 4]
```diff
@@ -129,22 +129,25 @@ void goto_trace_stept::output(
...
-static std::string
-numeric_representation(const exprt &expr, const trace_optionst &options)
+static std::string numeric_representation(
+  const constant_exprt &expr,
+  const trace_optionst &options)
 {
   std::string result;
   std::string prefix;
+
+  const irep_idt &value = expr.get_value();
```
Comment (tautschnig) [inline src/goto-programs/goto_trace.cpp:139]:
```suggestion
  const std::string &value = id2string(expr.get_value());
```
and remove the repeated `id2string` below.

File: src/util/arith_tools.cpp [hunk 5]
```diff
@@ -43,44 +43,57 @@ bool to_integer(const constant_exprt &expr, mp_integer &int_value)
...
+      int_value = bv2integer(id2string(value), width, true);
       return false;
     }
     else if(subtype.id()==ID_unsignedbv)
     {
-      int_value = bv2integer(id2string(value), false);
+      int_value = bv2integer(id2string(value), width, false);
+      return false;
+    }
+    else if(subtype.id()==ID_c_bool)
```
Comment (tautschnig) [inline src/util/arith_tools.cpp]: Dare I say that a bugfix without accompanying test has been sneaked in?

**Review by tautschnig — COMMENTED** [review]
A couple of suggestions (trying out GitHub's new feature...) below. From the start, this PR would have benefitted quite a bit from commit messages answering the "why?" question. The answer to that has become clear through other PRs, but that delays reviewing significantly.

File: src/ansi-c/expr2c.cpp (hunk 2 above)
Comment (kroening) [reply src/ansi-c/expr2c.cpp:1814]: Yes, but we appear to do some, and not others!

File: src/goto-programs/goto_convert.cpp (hunk 3 above)
Comment (kroening) [reply src/goto-programs/goto_convert.cpp]: Say here we add a 'const'.

File: src/goto-programs/goto_trace.cpp (hunk 4 above)
Comment (kroening) [reply src/goto-programs/goto_trace.cpp:139]: Uh, no, we are trying to change the signature of bv2integer to irep_idt.

File: src/util/arith_tools.cpp (hunk 5 above)
Comment (kroening) [reply src/util/arith_tools.cpp]: Will make separate PR.

File: src/ansi-c/expr2c.cpp (hunk 2 above)
Comment (tautschnig) [reply src/ansi-c/expr2c.cpp:1814]: As a reviewer, the first question I usually ask myself is "why was this line changed" and expect to be able to immediately answer that to myself from the commit message. Sometimes there are additional changes _in those lines_ beyond what's in the commit message. If those are of the nature "added in a `const`" or "fixed up whitespace" then those seem like desirable improvements. Everything else raises eyebrows and usually makes me post questions in here ;-)

File: src/goto-programs/goto_convert.cpp (hunk 3 above)
Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: I've tried to explain my thinking above, maybe that helps?

File: src/goto-programs/goto_convert.cpp [hunk 6]
```diff
@@ -1830,8 +1830,13 @@ bool goto_convertt::get_string_constant(
       forall_operands(it, index_op)
         if(it->is_constant())
         {
-          unsigned long i = integer2ulong(
-            bv2integer(id2string(to_constant_expr(*it).get_value()), true));
+          mp_integer int_value;
+          if(to_integer(to_constant_expr(*it), int_value))
+            return true;
```
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: The limitation of GitHub's suggestions feature (I have already provided feedback to them): my suggestion was meant to replace the entire bit of code, the lines above need to be removed now.

**Review by tautschnig — APPROVED** [review]
One further change required as below, but otherwise looks good to me.

File: src/goto-programs/goto_convert.cpp (hunk 6 above)
Comment (kroening) [reply src/goto-programs/goto_convert.cpp]: done.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 92f26219c50bd441f25de2932e491e7ec92be249).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88400326
